solidity/dashboard/src/services/token-staking.service.js

```diff
@@ -1,77 +1,75 @@
 import moment from "moment"
 import {
   isCodeValid,
   createManagedGrantContractInstance,
   CONTRACT_DEPLOY_BLOCK_NUMBER,
   ContractsLoaded,
+  Web3Loaded,
 } from "../contracts"
 import { isSameEthAddress } from "../utils/general.utils"
 import { isEmptyArray } from "../utils/array.utils"
+import { getEventsFromTransaction } from "../utils/ethereum.utils"
 
-const fetchDelegatedTokensData = async (web3Context) => {
-  const { yourAddress, grantContract, eth, web3 } = web3Context
-  const [
-    stakedBalance,
-    ownerAddress,
-    beneficiaryAddress,
-    authorizerAddress,
-    initializationPeriod,
-  ] = await Promise.all([
-    contractService.makeCall(
-      web3Context,
-      TOKEN_STAKING_CONTRACT_NAME,
-      "balanceOf",
-      yourAddress
-    ),
-    contractService.makeCall(
-      web3Context,
-      TOKEN_STAKING_CONTRACT_NAME,
-      "ownerOf",
-      yourAddress
-    ),
-    contractService.makeCall(
-      web3Context,
-      TOKEN_STAKING_CONTRACT_NAME,
-      "beneficiaryOf",
-      yourAddress
-    ),
-    contractService.makeCall(
-      web3Context,
-      TOKEN_STAKING_CONTRACT_NAME,
-      "authorizerOf",
-      yourAddress
-    ),
-    contractService.makeCall(
-      web3Context,
-      TOKEN_STAKING_CONTRACT_NAME,
-      "initializationPeriod"
-    ),
+const fetchDelegatedTokensData = async () => {
+  const contractsLoaded = await ContractsLoaded
+  const { grantContract, stakingContract } = contractsLoaded
+  const web3 = await Web3Loaded
+  const { eth } = web3
+  const { defaultAccount: yourAddress } = eth
+
+  let operatorStakedEvents
+  try {
+    operatorStakedEvents = await stakingContract.getPastEvents(
+      "OperatorStaked",
+      {
+        fromBlock: CONTRACT_DEPLOY_BLOCK_NUMBER.stakingContract,
+        filter: { operator: yourAddress },
+      }
+    )
+  } catch (err) {}
+
+  let [stakedBalance, ownerAddress, initializationPeriod] = await Promise.all([
+    stakingContract.methods.balanceOf(yourAddress).call(),
+    stakingContract.methods.ownerOf(yourAddress).call(),
+    stakingContract.methods.initializationPeriod().call(),
   ])
 
-  let isUndelegationFromGrant = true
-  let grantStakeDetails
+  let transactionHash
+  let beneficiaryAddress = ownerAddress
+  let authorizerAddress = ownerAddress
+
+  if (operatorStakedEvents.length > 0) {
+    ;({
+      transactionHash,
+      returnValues: {
+        beneficiary: beneficiaryAddress,
+        authorizer: authorizerAddress,
+      },
+    } = operatorStakedEvents[0])
+  }
+
+  const eventsToCheck = [[grantContract, "TokenGrantStaked"]]
+  let events
   try {
-    grantStakeDetails = await grantContract.methods
-      .getGrantStakeDetails(yourAddress)
-      .call()
-  } catch (error) {
-    isUndelegationFromGrant = false
+    events = await getEventsFromTransaction(eventsToCheck, transactionHash)
+  } catch (err) {}
+
+  let isDelegationFromGrant = true
+  let grantId
+  if (events && events.TokenGrantStaked) {
```

Comment:
I think we can handle it in a single `if` statement.

```
let isDelegationFromGrant = false
let isManagedGrant = false
let managedGrantContractInstance
let grantId
if (events && events.TokenGrantStaked) {
    grantId = events.TokenGrantStaked.grantId
    isDelegationFromGrant = true
    const { grantee } = await grantContract.methods.getGrant(grantId).call()
    // check if grantee is a contract
    const code = await eth.getCode(grantee)
    ownerAddress = grantee
    if (isCodeValid(code)) {
      managedGrantContractInstance = createManagedGrantContractInstance(
        web3,
        grantee
      )
      isManagedGrant = true
      ownerAddress = await managedGrantContractInstance.methods.grantee().call()
    }
} else {
    ownerAddress = stakingContract.methods.ownerOf(yourAddress).call()
}
```

Thoughts?

Reply:
Good idea, I will definitely implement that.